templates/3b_Generic_Cluster_Cells.ipynb

```diff
@@ -0,0 +1,665 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #8.    cell_som_input_data = pd.read_csv(cell_som_input_name)
</pre><p>We should either put the file I/O in a separate step where the user can specify the extension, or abstract it within the train_som function so that it can support CSV files directly</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/892/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Reply:
We can definitely include support for both CSV and feather files, not a huge fan though because it becomes quite bit messy to do so (we have several places where we read and write to the cell SOM input file, currently assumed to always be `.feather`).

I like option 1 the most, where the user specifies the extension in the notebook. I still think we should do the `.csv` to `.feather` conversion in the notebook, because then we can also reset the extension of the cell SOM input file across all functions to `.feather` as well (rather than have to handle it on a case-by-case basis in each function).

All this will be much easier to handle we convert the entire process to a class-based system.

Reply:
What about avoiding any writes until the end of the pipeline? Will that require rewriting a ton of stuff? It seems like loading a df, appending to it, and then at the end saving would be simpler, given that the cell table is never going to be so big that we need to read/write parts of it at a time

Reply:
It shouldn't be too bad to implement, especially since we pre-compute `cluster_counts_size_norm` in notebook 3 (the inputs for SOM training). I'll experiment with this a bit today.

Reply:
I'm going to just do this for the SOM input training for now (`cluster_counts_size_norm` in notebook 3, and `cell_som_input_data` in notebook 3b), since this is the only data that will be significantly changed for both SOM clustering, consensus clustering, and remapping.